### ISSUE TITLE
[RFC] Open source RayCommonLibs, which'll enrich the Ray ecosystem. 

### ISSUE THREAD

**Issue by jovany-wang** [ray-project/ray#20760]
An initial public proposal doc: https://docs.google.com/document/d/1keQjKNKmhlC7SJQCGByPB-knnVMk9W_0khh7qpZv9Js/edit#heading=h.k3w9s2fjmhwo

Some related issues:
[[RFC] Feature request about Actor State/Migration Notification. ](https://github.com/ray-project/ray/issues/13833)

-------Updated
An pubsub lightweight library proposal: https://docs.google.com/document/d/1eOiTBDO8Crpc4GgKKrKuy-IlUYzKkBrYevopXA8ksxA/edit#heading=h.y9vrwvnhp4t4

**Comment by jovany-wang** [comment]
@ericl @simon-mo @raulchen @lixin-wei CC

**Comment by ericl** [comment]
This seems quite similar to the actor pool proposal from @edoakes for improving actors management for Serve. If we go ahead with the ActorPool concept, it seems these efforts can be unified.

**Comment by simon-mo** [comment]
Are the libraries implemented purely in Java at the moment?

**Comment by jovany-wang** [comment]
> Are the libraries implemented purely in Java at the moment?

Some of them are implemented in multiple languages. The ActorStateNotification is implemented in Java currently.

**Comment by jovany-wang** [comment]
@simon-mo Any concern on introducing a new language for RayServe?

**Comment by simon-mo** [comment]
Serve and other Ray Python libraries cannot depend on Java due to memory overhead and JVM dependencies. So this has to be either Python, C++ actor, or C++ core component.

**Comment by jovany-wang** [comment]
> Serve and other Ray Python libraries cannot depend on Java due to memory overhead and JVM dependencies. So this has to be either Python, C++ actor, or C++ core component.

I see.

**Comment by qicosmos** [comment]
> https://docs.google.com/document/d/1keQjKNKmhlC7SJQCGByPB-knnVMk9W_0khh7qpZv9Js/edit#heading=h.k3w9s2fjmhwo

I think we can use c++ worker to implement the RayCommonLibs according to the proposal.

**Comment by jovany-wang** [comment]
@ericl @edoakes @simon-mo Is there any doc or related issue around actor pool concept?

**Comment by jovany-wang** [comment]
Hi @ericl and @raulchen ,

I had talked with @simon-mo and @liuyang-my about the `ray.util.EventBroker` and `ray.util.ActorStateObserver`. We've got phased agreements on the following points:
 - We both think`ray.util.EventBroker` and  `ray.util.ActorStateObserver` are necessary requirements for RayServe/RayServing. (Also it's very useful for other Ray applications including other apps from Ant.)
 -  Backend implementations of the actors will be write in Python instead of Java. And multiple languages APIs will be supported by a thin wrapper.
 - We make them as a part of `ray.util` package currently, like `ray.util.actor_group`.

We're making efforts on the above items, such as Simon is making efforts on API review and I'm rewriting it in Python, and we'll put them into `ray.util` package.

But for a long term plan, I'm very open to discuss the topic that whether `ray.util` is the best place for these components.
As described in the title of this issue, we'd like to propose `commonlibs` library for Ray ecosystem to bridge Ray applications and Ray core. The key difference between `commonlibs` and `ray.util` package is `commonlibs` is an independent package/library. This is very used for hot bug fix in practice.
